File: src/download_page.py
from PySide6.QtCore import QObject, Qt
from PySide6.QtWidgets import QLabel, QFrame, QVBoxLayout, QHBoxLayout, QPushButton, QSpacerItem, QSizePolicy
from src.thread_manager import thread_manager
from src.downloader.downloader import DownloadTask


class DownloadPageManager(QObject):
    """下载页面管理器"""
    def __init__(self, download_ui):
        super().__init__()
        self.download_ui = download_ui
        self.download_tasks = {}
        thread_manager.pikpak_events_loaded.connect(self.display_events)
        thread_manager.pikpak_download_url_loaded.connect(self.handle_download_url)
        self.setup_download_list()

    def setup_download_list(self):
        """设置下载列表UI"""
        layout = self.download_ui.verticalLayout_3
        layout.setSpacing(5)
        layout.setContentsMargins(5, 5, 5, 5)
        layout.addSpacerItem(QSpacerItem(20, 40, QSizePolicy.Policy.Minimum, QSizePolicy.Policy.Expanding))

    @staticmethod
    def load_recent_files():
        """加载最近添加的文件列表"""
        thread_manager.fetch_pikpak_events()

    def display_events(self, events):
        """在UI中显示事件列表"""
        layout = self.download_ui.verticalLayout_2
        while layout.count():
            layout.takeAt(0).widget().deleteLater()
        if not events:
            label = QLabel("暂无最近添加的文件")
            label.setAlignment(Qt.AlignmentFlag.AlignCenter)
            label.setStyleSheet("font-size: 14px; color: #999; padding: 20px")
            layout.addWidget(label)
        for event in events:
            self._create_event_item(event, layout)
        layout.addStretch()

    def _create_event_item(self, event, parent_layout):
        """创建单个事件项"""
        ref_resource = event.get('reference_resource', {})
        file_name = event.get('file_name', '')
        # 框架
        item = QFrame()
        item.setFixedHeight(60)
        item.setStyleSheet("QFrame {border: 1px solid #e0e0e0}"
                           "QFrame:hover {background-color: #f0f0f0}")
        # 事件数据
        item.event_data = event
        # 点击事件
        item.mousePressEvent = lambda e, clicked_item=item: self._on_item_clicked(clicked_item)
        item_layout = QVBoxLayout(item)
        item_layout.setContentsMargins(10, 5, 10, 5)
        # 文件名
        name_label = QLabel(file_name)
        name_label.setStyleSheet("font-size: 14px; background-color: transparent; border: none")
        name_label.setWordWrap(True)
        item_layout.addWidget(name_label)
        # 类型和大小
        type_label = QLabel(self._get_type_size_text(ref_resource))
        type_label.setStyleSheet("font-size: 12px; background-color: transparent; border: none")
        item_layout.addWidget(type_label)
        parent_layout.addWidget(item)

    @staticmethod
    def _on_item_clicked(item):
        """点击事件"""
        ref_resource = item.event_data.get('reference_resource', {})
        if ref_resource.get('kind') == 'drive#folder':
            return
        file_id = ref_resource.get('id')
        thread_manager.fetch_pikpak_download_url(file_id)

    def handle_download_url(self, result):
        """处理获取到的下载链接"""
        if not result:
            return
        download_url = self._extract_download_url(result)
        if download_url:
            file_name = result.get('name', '')
            if not file_name:
                file_name = download_url.split('/')[-1].split('?')[0] or "download.file"
            self.add_download_task(download_url, file_name)

    @staticmethod
    def _extract_download_url(result):
        """从结果中提取下载链接"""
        for media in result.get('medias', []):
            if media.get('link', {}).get('url'):
                return media['link']['url']
        if result.get('web_content_link'):
            return result['web_content_link']
        for link_data in result.get('links', {}).values():
            if link_data.get('url'):
                return link_data['url']
        return None

    def add_download_task(self, url, file_name):
        """添加下载任务并显示在UI中"""
        task = DownloadTask(url, file_name)
        self.download_tasks[file_name] = task
        task_widget = self._create_download_item(file_name)
        task.progress_changed.connect(lambda progress, downloaded, total, w=task_widget: self._update_download_progress(w, progress, downloaded, total))
        task.status_changed.connect(lambda status, w=task_widget: self._update_download_status(w, status))
        layout = self.download_ui.verticalLayout_3
        if layout.count() > 0:
            layout.takeAt(layout.count() - 1)
        layout.addWidget(task_widget)
        layout.addSpacerItem(QSpacerItem(20, 40, QSizePolicy.Policy.Minimum, QSizePolicy.Policy.Expanding))
        task.start()

    def _create_download_item(self, file_name):
        """创建下载任务UI项"""
        widget = QFrame()
        widget.setFixedHeight(50)
        widget.setStyleSheet("QFrame {border: 2px solid #9E9E9E; border-radius: 4px; background-color: transparent}")
        # 创建进度背景
        progress_frame = QFrame(widget)
        progress_frame.setGeometry(0, 0, 0, 50)
        progress_frame.setStyleSheet("background-color: rgba(76, 175, 80, 0.3);border: none; border-radius: 3px")
        progress_frame.lower()
        # 内容布局
        layout = QHBoxLayout(widget)
        layout.setContentsMargins(10, 5, 10, 5)
        # 状态
        status_label = QLabel("等待")
        status_label.setStyleSheet("font-size: 13px; font-weight: bold; background-color: transparent; border: none; min-width: 40px")
        status_label.setFixedWidth(40)
        layout.addWidget(status_label)
        # 文件名
        name_label = QLabel(file_name)
        name_label.setStyleSheet("font-size: 13px; font-weight: bold; background-color: transparent; border: none")
        name_label.setWordWrap(True)
        layout.addWidget(name_label, 1)
        # 进度文字
        progress_label = QLabel("0B/0B")
        progress_label.setStyleSheet("font-size: 13px; font-weight: bold; background-color: transparent; border: none")
        layout.addWidget(progress_label)
        # 删除按钮
        remove_btn = QPushButton("删除")
        remove_btn.setFixedSize(50, 25)
        remove_btn.setStyleSheet("QPushButton {font-size: 11px; color: #ff6b6b; background-color: transparent; border: 1px solid #ff6b6b; border-radius: 3px}"
                                 "QPushButton:hover {background-color: rgba(255, 107, 107, 0.1)}")
        remove_btn.clicked.connect(lambda: self._remove_task(widget))
        layout.addWidget(remove_btn)
        # 保存引用
        widget.file_name = file_name
        widget.progress_frame = progress_frame
        widget.status_label = status_label
        widget.name_label = name_label
        widget.progress_label = progress_label
        widget.remove_btn = remove_btn
        # 点击暂停/继续
        widget.mousePressEvent = lambda e, w=widget: self._toggle_pause(w)
        return widget

    def _update_download_progress(self, widget, progress, downloaded, total):
        """更新下载进度"""
        if progress > 0:
            progress_width = int(widget.width() * progress / 100)
            widget.progress_frame.setFixedWidth(progress_width)
        downloaded_str = self._format_size(downloaded)
        total_str = self._format_size(total)
        widget.progress_label.setText(f"{downloaded_str}/{total_str}")

    @staticmethod
    def _update_download_status(widget, status):
        """更新下载状态"""
        widget.status_label.setText(status)
        color_map = {"下载中": "#4CAF50", "已暂停": "#FF9800", "完成": "#2196F3", "错误": "#F44336", "等待": "#9E9E9E"}
        border_color = color_map.get(status, "#9E9E9E")
        widget.setStyleSheet(f"QFrame {{border: 2px solid {border_color}; border-radius: 4px; background-color: transparent}}")

    def _toggle_pause(self, widget):
        """切换暂停/继续"""
        file_name = widget.file_name
        if file_name in self.download_tasks:
            task = self.download_tasks[file_name]
            if task.status == "下载中":
                task.pause()
            elif task.status == "已暂停":
                task.status = "下载中"
                task.is_running = True
                task.start()

    def _remove_task(self, widget):
        """删除下载任务"""
        file_name = widget.file_name
        if file_name in self.download_tasks:
            task = self.download_tasks[file_name]
            task.stop()
            del self.download_tasks[file_name]
        widget.setParent(None)
        widget.deleteLater()

    @staticmethod
    def _format_size(size):
        """格式化文件大小"""
        for unit in ['B', 'KB', 'MB', 'GB']:
            if size < 1024.0:
                return f"{size:.1f}{unit}"
            size /= 1024.0
        return f"{size:.1f}TB"

    @staticmethod
    def _get_type_size_text(ref_resource):
        """获取类型和大小"""
        if ref_resource.get('kind') == 'drive#folder':
            return "文件夹"
        mime_type = ref_resource.get('mime_type', '')
        size = ref_resource.get('size', '0')
        file_type = "视频" if 'video' in mime_type else "文件"
        if size.isdigit():
            size_int = int(size)
            return f"{file_type} {DownloadPageManager._format_size(size_int)}"
        return f"{file_type} 0B"

File: windows/downloader/download_page.cpp
#include "download_page.h"
#include <QLabel>
#include <QPointer>
#include <QJsonArray>
#include <QJsonObject>
#include <QFileDialog>
#include "../config.h"
#include "chunk_download.h"
#include "../api/pikpak_api.h"

DownloadPage::DownloadPage(QWidget *parent): QWidget(parent)
{
    ui.setupUi(this);
    applyTheme();
    setupDownloadPathUI();
    setupDownloadList();
}

void DownloadPage::setManagers(PikPakApi *pikpakapi)
{   // 初始化实例
    pikpakApi = pikpakapi;
}

void DownloadPage::applyTheme() const
{   // 主题
    const QColor color1 = getConfig("Theme/color1", 0xfdf7ff).toString();
    ui.mainFrame->setStyleSheet(QString("QFrame {background-color: %1}").arg(color1.name()));
}

void DownloadPage::setupDownloadPathUI()
{   // 下载路径
    downloadPath = getConfig("Download/download_path").toString();
    ui.lineEditDownloadPath->setText(downloadPath);
    connect(ui.btnSelectPath, &QPushButton::clicked, this, &DownloadPage::selectDownloadPath);
}

void DownloadPage::selectDownloadPath()
{   // 设置下载路径
    const QString currentPath = ui.lineEditDownloadPath->text();
    if (const QString dirPath = QFileDialog::getExistingDirectory(this, "选择下载路径", currentPath); !dirPath.isEmpty()) {
        ui.lineEditDownloadPath->setText(dirPath);
        setConfig("Download/download_path", dirPath);
        downloadPath = dirPath;
    }
}

void DownloadPage::displayItems(const QJsonArray &items, const QString &emptyMessage)
{   // 显示组件
    clearLayout(ui.fileListLayout);
    if (items.isEmpty()) {
        auto *label = new QLabel(emptyMessage);
        label->setAlignment(Qt::AlignCenter);
        label->setStyleSheet("font-size: 14px; color: #999; padding: 20px");
        ui.fileListLayout->addWidget(label);
        ui.fileListLayout->addStretch();
        return;
    }
    for (const auto &val : items) {
        QJsonObject obj = val.toObject();
        QString typeText = obj["kind"].toString() == "drive#folder" ? "文件夹" : "文件";
        QString sizeText;
        if (typeText == "文件") {
            const qint64 size = obj["size"].toString().toLongLong();
            sizeText = " " + QLocale().formattedDataSize(size);
        }
        auto *item = new QFrame(this);
        item->setStyleSheet("QFrame {border: 1px solid #e0e0e0}"
                            "QFrame:hover {background-color: #f0f0f0}");
        item->setProperty("type", typeText);
        item->setProperty("file_id", obj["id"].toString());
        item->setProperty("file_name", obj["name"].toString());
        item->installEventFilter(this);
        auto *layout = new QVBoxLayout(item);
        auto *nameLabel = new QLabel(obj["name"].toString());
        nameLabel->setStyleSheet("font-size: 14px; background: transparent; border: none");
        nameLabel->setWordWrap(true);
        nameLabel->setAttribute(Qt::WA_TransparentForMouseEvents);
        layout->addWidget(nameLabel);
        auto *typeLabel = new QLabel(typeText + sizeText);
        typeLabel->setStyleSheet("font-size: 12px; background: transparent; border: none");
        typeLabel->setAttribute(Qt::WA_TransparentForMouseEvents);
        layout->addWidget(typeLabel);
        ui.fileListLayout->addWidget(item);
    }
    ui.fileListLayout->addStretch();
}

void DownloadPage::loadRecentFiles()
{   // 显示最近添加
    QPointer guard(this);
    pikpakApi->getRecentFiles(3, [guard](const QJsonObject &result, const int statusCode, const QString &error) {
        if (!guard) return;
        if (statusCode == 200 && error.isEmpty()) {
            QJsonArray events = result["events"].toArray();
            QJsonArray items;
            for (const auto &val : events) {
                QJsonObject event = val.toObject();
                QJsonObject ref = event["reference_resource"].toObject();
                QJsonObject item;
                item["id"] = event["file_id"].toString();
                item["name"] = event["file_name"].toString();
                item["kind"] = ref["kind"].toString();
                item["size"] = ref["size"].toString();
                items.append(item);
            }
            guard->displayItems(items, "暂无最近添加的文件");
        } else guard->displayItems(QJsonArray(), "获取最近文件失败：" + error);
    });
}

void DownloadPage::fetchAndDownload(const QString &fileId, const QString &fileName)
{   // 获取链接并下载
    QFrame *widget = taskWidgets.value(fileName, nullptr);
    updateDownloadStatus(widget, "获取链接");
    const QPointer guard(this);
    pikpakApi->getDownloadUrl(fileId, 3, [=](const QJsonObject &fileInfo, const int statusCode, const QString &error) {
        if (!guard) return;
        if (statusCode != 200 || !error.isEmpty()) {
            updateDownloadStatus(widget, "获取链接失败");
            return;
        }
        QString downloadUrl;
        const auto medias = fileInfo["medias"].toArray();
        if (!medias.isEmpty()) {
            auto link = medias.first().toObject()["link"].toObject();
            auto mirrors = link["mirrors"].toArray();
            if (!mirrors.isEmpty()) downloadUrl = mirrors[mirrors.size() >= 2 ? 1 : 0].toString();
            else downloadUrl = link["url"].toString();
        }
        if (downloadUrl.isEmpty()) downloadUrl = fileInfo["web_content_link"].toString();
        const QString savePath = downloadPath + "/" + fileName;
        auto *task = new ChunkDownload(downloadUrl, savePath, 4, this);
        downloadTasks[fileName] = task;
        connect(task, &ChunkDownload::progressChanged, this, [this, fileName](const int percent, const qint64 downloaded, const qint64 total) {
            if (const QFrame *w = taskWidgets.value(fileName, nullptr)) updateDownloadProgress(w, percent, downloaded, total);});
        connect(task, &ChunkDownload::statusChanged, this, [this, fileName](const QString &status) {
            if (QFrame *w = taskWidgets.value(fileName, nullptr)) updateDownloadStatus(w, status);});
        task->start();
    });
}

void DownloadPage::loadFolderContent(const QString &folderId)
{   // 显示文件夹内容
    QPointer guard(this);
    pikpakApi->getFileList(folderId, 100, "", 3, [guard, folderId](const QJsonObject &result, const int statusCode, const QString &error) {
        if (!guard) return;
        if (statusCode == 200 && error.isEmpty()) {
            const QJsonArray files = result["files"].toArray();
            guard->displayItems(files, "文件夹为空");
            guard->currentFolderId = folderId;
        } else guard->displayItems(QJsonArray(), "获取文件夹内容失败：" + (error.isEmpty() ? QString::number(statusCode) : error));
    });
}

bool DownloadPage::eventFilter(QObject *obj, QEvent *event)
{   // 点击事件
    if (event->type() == QEvent::MouseButtonPress) {
        const auto *frame = qobject_cast<QFrame*>(obj);
        if (!frame) return QWidget::eventFilter(obj, event);
        const QString type = frame->property("type").toString();
        const QString fileId = frame->property("file_id").toString();
        const QString fileName = frame->property("file_name").toString();
        if (type == "文件") {
            if (taskWidgets.contains(fileName) || downloadTasks.contains(fileName)) return true;
            QFrame *widget = createDownloadItem(fileName);
            taskWidgets[fileName] = widget;
            ui.taskListLayout->insertWidget(0, widget);
            fetchAndDownload(fileId, fileName);
            return true;
        }
        loadFolderContent(fileId);
        return true;
    }
    return QWidget::eventFilter(obj, event);
}

void DownloadPage::setupDownloadList() const
{   // 设置下载列表
    ui.taskListLayout->setSpacing(5);
    ui.taskListLayout->setContentsMargins(5, 5, 5, 5);
    ui.taskListLayout->addSpacerItem(new QSpacerItem(20, 40, QSizePolicy::Minimum, QSizePolicy::Expanding));
}

QFrame* DownloadPage::createDownloadItem(const QString &fileName)
{   // 创建下载组件
    auto *widget = new QFrame(this);
    widget->setFixedHeight(50);
    widget->setStyleSheet("QFrame {border: 2px solid #9E9E9E; border-radius: 4px; background: transparent}");
    auto *progressFrame = new QFrame(widget);
    progressFrame->setGeometry(0, 0, 0, 50);
    progressFrame->setStyleSheet("background-color: rgba(76, 175, 80, 0.3); border: none; border-radius: 3px");
    progressFrame->lower();
    auto *layout = new QHBoxLayout(widget);
    layout->setContentsMargins(10, 5, 10, 5);
    auto *statusLabel = new QLabel("获取链接");
    statusLabel->setStyleSheet("font-size: 13px; font-weight: bold; background: transparent; border: none; min-width: 40px");
    layout->addWidget(statusLabel);
    auto *nameLabel = new QLabel(fileName);
    nameLabel->setStyleSheet("font-size: 13px; font-weight: bold; background: transparent; border: none");
    nameLabel->setWordWrap(true);
    layout->addWidget(nameLabel, 1);
    auto *progressLabel = new QLabel("0B/0B");
    progressLabel->setStyleSheet("font-size: 13px; font-weight: bold; background: transparent; border: none");
    layout->addWidget(progressLabel);
    auto *removeBtn = new QPushButton("删除");
    removeBtn->setFixedSize(50, 25);
    removeBtn->setStyleSheet("QPushButton {font-size: 11px; color: #ff6b6b; background: transparent; border: 1px solid #ff6b6b; border-radius: 3px}"
                             "QPushButton:hover {background: rgba(255, 107, 107, 0.1)}");
    layout->addWidget(removeBtn);
    widget->setProperty("file_name", fileName);
    widget->setProperty("progress_frame", QVariant::fromValue(progressFrame));
    widget->setProperty("status_label", QVariant::fromValue(statusLabel));
    widget->setProperty("progress_label", QVariant::fromValue(progressLabel));
    widget->installEventFilter(this);
    connect(removeBtn, &QPushButton::clicked, this, [this, fileName] {
        ChunkDownload *task = downloadTasks.take(fileName);
        task->stop();
        task->deleteLater();
        taskWidgets.take(fileName)->deleteLater();
    });
    return widget;
}

void DownloadPage::updateDownloadProgress(const QFrame *widget, const int percent, const qint64 downloaded, const qint64 total)
{   // 更新下载进度
    const int width = widget->width() * percent / 100;
    widget->property("progress_frame").value<QFrame*>()->setFixedWidth(width);
    QString downloadedStr = QLocale().formattedDataSize(downloaded);
    QString totalStr = QLocale().formattedDataSize(total);
    widget->property("progress_label").value<QLabel*>()->setText(QString("%1/%2").arg(downloadedStr, totalStr));
}

void DownloadPage::updateDownloadStatus(QFrame *widget, const QString &status)
{   // 更新下载状态
    widget->property("status_label").value<QLabel*>()->setText(status);
    const QMap<QString, QString> colorMap = {
        {"下载中", "#4CAF50"},
        {"完成", "#2196F3"},
        {"错误", "#F44336"},
        {"获取链接", "#FF9800"}
    };
    const QString color = colorMap.value(status, "#9E9E9E");
    widget->setStyleSheet(QString("QFrame {border: 2px solid %1; border-radius: 4px; background: transparent}").arg(color));
}

void DownloadPage::clearLayout(QLayout *layout)
{   // 清理组件
    while (const QLayoutItem *item = layout->takeAt(0)) {
        if (QWidget *w = item->widget()) w->deleteLater();
        delete item;
    }
}